src/AspNetCoreModuleV2/RequestHandlerLib/filewatcher.cpp

```diff
@@ -354,22 +354,29 @@ FILE_WATCHER_ENTRY::Monitor(VOID)
 {
     HRESULT hr = S_OK;
     DWORD   cbRead;
-
     AcquireSRWLockExclusive(&_srwLock);
     ReferenceFileWatcherEntry();
     ZeroMemory(&_overlapped, sizeof(_overlapped));
 
-    if (!ReadDirectoryChangesW(_hDirectory,
-        _buffDirectoryChanges.QueryPtr(),
-        _buffDirectoryChanges.QuerySize(),
-        FALSE,        // Watching sub dirs. Set to False now as only monitoring app_offline
-        FILE_NOTIFY_VALID_MASK & ~FILE_NOTIFY_CHANGE_LAST_ACCESS,
-        &cbRead,
-        &_overlapped,
-        NULL))
+    // Check if file exist before polling for changes
```

Comment:
Why is this check necessary?

Reply:
If app_offline exists at the time we start file watcher it would not get reported an application would not be stopped. This is a fix for a race that `AppOfflineDroppedWhileSiteStarting_SiteShutsDown_InProcess` is trying to force.

Reply:
This may reduce the size of the race window, but isn't there still a race?  What if app_offline is added between the calls to `GetFileAttributes` and `ReadDirectoryChangesW`?

Reply:
There is an inherent race when using file watcher, the file might have appeared between we called the method and OS initialized data structures required for while watching. This at least takes race window from hundreds of milliseconds to tens of milliseconds. 

Let me think if moving exists check after ReadDirectoryChangesW call would help.

Reply:
I would expect there to be some way to check for the existence of a file without any races.  Worst case you could always fallback to polling (say once per second) in case everything else fails.  But it might also be sufficient to check for existence exactly once after the file watcher is created.